server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -387,43 +402,41 @@ boolean isChannelDisconnected() {
          return !ch.isOpen();
       }
 
-      void sendEvent(byte[] key, byte[] value, long dataVersion, CacheEntryEvent event) {
+      CompletionStage<Void> sendEvent(byte[] key, byte[] value, long dataVersion, CacheEntryEvent event) {
          Events.Event remoteEvent = createRemoteEvent(key, value, dataVersion, event);
          if (isTrace)
             log.tracef("Queue event %s, before queuing event queue size is %d", remoteEvent, eventQueue.size());
 
          EventLoop loop = ch.eventLoop();
-         // We shouldn't be in the event loop, but just in case we can't get stuck blocking on putting into the queue
-         // so we offer and try to catch up on events if possible
-         if (loop.inEventLoop()) {
-            boolean offered = eventQueue.offer(remoteEvent);
-            while (!offered) {
-               // If the event queue is full, we have to try to write some events to free up space since we are in the
-               // event loop and no other thread can drain this queue but us
-               writeEventsIfPossible();
-               // We again try to offer, but if we weren't able to write any events this will not offer - so we
-               // have to wait for the client to catch up to us - we put a little wait to not cause CPU to spin
-               try {
-                  offered = eventQueue.offer(remoteEvent, 1, TimeUnit.MILLISECONDS);
-               } catch (InterruptedException e) {
-                  throw new CacheException(e);
-               }
+         int size = eventSize.incrementAndGet();
+         boolean forceWait = size >= highWaterMark;
+         CompletableFuture<Void> cf;
+         if (forceWait) {
+            if (isTrace) {
+               log.tracef("Pending event size is %s which is forcing %s to delay operation until it is sent", size, event);
             }
-         } else {
-            try {
-               // TODO: replace with a better number
-               if (!eventQueue.offer(remoteEvent, 30, TimeUnit.SECONDS)) {
-                  throw new TimeoutException("Timed out attempting to offer remote event into queue");
-               }
-            } catch (InterruptedException e) {
-               throw new CacheException(e);
+            cf = new CompletableFuture<>();
+            // Have to write to map before queue, as we check in opposite order when sending events
+            delayedEvents.add(cf);
+
+            // If the event size is now less than water mark then don't force block - this is if there is a concurrent
+            // thread freeing all the waiters - It is possible but very slight chance that it could exhaust the
+            // entire delayedEvents without seeing ours so we don't force it to block (could get stuck otherwise)
```

Comment:
I think I understand the problem you're trying to fix, but I don't like the explanation: I would have liked a direct reference to "unblockCommands" instead of the completely different "freeing all the waiters", and the same with  "force it to block" vs "forceWait".

I'm also starting to think that the blocking should be different: keep a single queue and add a `CompletableFuture` field in `Events.Event`, populate it only if the queue size is above the (high?) watermark, and complete the future (if non-null) exactly when the event is written to the channel.

Reply:
I thought about something similar, but have to admit this idea is a bit cleaner. However the more I think on it I am not sold on running this on the netty thread.

Reply:
@wburns you mean it's not safe to unblock events on the netty thread because that might invoke some long running method, or even block if it's supposed to run on an executor and that executor is full?

Reply:
It is more just that we have had issues in the past of running things on the netty thread which has bitten us. I have tried to stay away from running things on there whenever possible. The more I think about it the completion should really be done on the cpu thread pool anyways and this one in the class here should be removed.

Reply:
So I think putting the `CompleteableFuture` in the `Events.Event` is good and then submit the completion to the cpu thread pool when completed would be a good solution personally.

Reply:
Although I don't know if netty threads would eventually become cpu threads as well :) Either way this can be explored in the JIRA created.

Reply:
I agree it would be best to run anything that's not strictly needed for writing to the channel on the cpu thread pool. The threading model of Netty means threads are not interchangeable, so they will always have to be a separate thread pool IMO.